DB/src/main/java/io/deephaven/db/v2/locations/local/FileTableLocationKey.java

```diff
@@ -0,0 +1,93 @@
+import io.deephaven.db.v2.locations.PartitionedTableLocationKey;
+import io.deephaven.db.v2.locations.TableLocationKey;
+import io.deephaven.io.log.impl.LogOutputStringImpl;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.util.Map;
+
+/**
+ * Base {@link ImmutableTableLocationKey} implementation for table locations that may be enclosed by partitions and
+ * described by a {@link File}.
+ * Sub-classes should override {@link #compareTo(TableLocationKey)} and {@link #equals(Object)} only if they need to
+ * prevent equality with other {@link FileTableLocationKey} implementations.
+ */
+public abstract class FileTableLocationKey extends PartitionedTableLocationKey {
+
+    protected final File file;
+
+    private int cachedHashCode;
+
+    /**
+     * Construct a new FileTableLocationKey for the supplied {@code file} and {@code partitions}.
+     *
+     * @param file        The file (or directory) that backs the keyed location. Will be adjusted to an absolute path.
+     * @param partitions The table partitions enclosing the table location keyed by {@code this}. Note that if this
+     *                   parameter is {@code null}, the location will be a member of no partitions. An ordered copy
+     *                   of the map will be made, so the calling code is free to mutate the map after this call
+     *                   completes, but the partition keys and values themselves <em>must</em> be effectively immutable.
+     */
+    public FileTableLocationKey(@NotNull final File file, @Nullable final Map<String, Comparable<?>> partitions) {
+        super(partitions);
+        this.file = file.getAbsoluteFile();
+    }
+
+    public final File getFile() {
+        return file;
+    }
+
+    @Override
+    public LogOutput append(@NotNull final LogOutput logOutput) {
+        return logOutput.append(getImplementationName())
+                .append(":[file=").append(file.getPath())
+                .append(",partitions=").append(PartitionsFormatter.INSTANCE, partitions)
+                .append(']');
+    }
+
+    @Override
+    public String toString() {
+        return new LogOutputStringImpl().append(this).toString();
+    }
+
+    @Override
+    public int compareTo(@NotNull final TableLocationKey other) {
+        if (other instanceof FileTableLocationKey) {
+            final FileTableLocationKey otherTyped = (FileTableLocationKey) other;
+            final int partitionComparisonResult = PartitionsComparator.INSTANCE.compare(partitions, otherTyped.partitions);
+            return partitionComparisonResult != 0 ? partitionComparisonResult : file.compareTo(otherTyped.file);
+        }
+        throw new ClassCastException("Cannot compare " + getClass() + " to " + other.getClass());
+    }
+
+    @Override
+    public int hashCode() {
+        if (cachedHashCode == 0) {
+            final int computedHashCode = 31 * partitions.hashCode() + file.hashCode();
```

Comment:
Getting the hashCode for a LinkedHashMap is pretty expensive, it iterates over all the elements.   Do we really need that?  The hashCode of the file is based on the file of its path with some shuffle added, wouldn't that be plenty for our purposes here, since the same path is "unlikely" [TM] to be in more than on partition?

Reply:
I don't want to change this. I *am* caching the hash code, on the bright side.
Imagine a scenario where partitions aren't part of the file path.
Or where we don't even have a file path, but rather some opaque URL.